core/src/main/java/com/linecorp/armeria/server/RoutingTrie.java

```diff
@@ -437,7 +459,12 @@ private void addRoute(Node<V> node, String path, V value) {
 
         // These values are sorted every time a new value is added.
         @Nullable
-        private List<V> values;
+        private List<V> valuesWithHighPrecedence;
+        @Nullable
+        private List<V> valuesWithLowPrecedence;
+        // Lazily populated from `valuesWithHighPrecedence` and `valuesWithLowPrecedence`.
```

Comment:
If I'm not mistaken, the mutations all happen from a `Builder` of the router. Is it possible to split a `NodeBuilder` and a built `Node` which only has an immutable `allValues`? I think it would be simpler code than having to maintain an `allValues` based on two other lists.

Alternatively, since a router is mostly a singleton, I'm not sure we need to lazy initialize these lists, and it would make it much easier to read without the lazy calls.